### PR TITLE
Improvements to animation keyframe computation

### PR DESCRIPTION
**Pull request by mrobinson**
This pull request contains two changes:

**Caching computed keyframes between animation changes**:
Instead of recomputing keyframe data for every tick of an animation,
cache the computed values when animations change. In addition to being
more efficient, this will allow us to return animation rules as property
declarations because we don't need to consult the final style to produce
them.

**Better computation of keyframe data**:
Instead of naively using `apply_declarations` to compute the style of
each keyframe, use an approach more like Gecko's where we carefully
walk through the animation keyframes and try to extract `AnimationValue`s
for each computed keyframe. In this approach we respect the order with
which the properties are set in the keyframe source and try to deal with
CSS custom properties.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes.

### CODE REVIEW

**Comment by mrobinson** [comment]
I've added a closely related change to this PR which improves the computation of keyframe data and ditches Servo's use of `apply_declarations` for this.

**Comment by mrobinson** [comment]
This needs a little bit more work.

**Comment by jdm** [comment]
I read through 917350369e6251b4296d658f2ce388193d000603 and it looks sensible.

**Comment by mrobinson** [comment]
@jdm Thanks for looking at the first commit. That should be unchanged in the newest version of the branch. The second commit is unfortunately a bit large, which means it's probably harder to review. I wasn't sure how to split it into simpler chunks. 😕 I'm happy to walk through it with anyone or do whatever else is necessary to make it easier to review.

**Review by jdm — APPROVED** [review]
These changes look sensible to me.

File: components/style/animation.rs [hunk 1]
```diff
@@ -172,100 +173,223 @@ pub enum KeyframesIterationState {
...
+        base_style: &Arc<ComputedValues>,
+    ) {
+        // Each keyframe declaration may optionally specify a timing function, falling
+        // back to the one defined global for the animation.
+        let guard = &context.guards.author;
+        if let Some(timing_function) = step.get_animation_timing_function(&guard) {
+            self.timing_function = Some(timing_function.to_computed_value_without_context());
+        }
+
+        match step.value {
```
Comment (emilio) [inline components/style/animation.rs:236]: Maybe `let block = match setep.value { ... };`, then de-indent the rest?

File: components/style/animation.rs [hunk 2]
```diff
@@ -172,100 +173,223 @@ pub enum KeyframesIterationState {
...
+        debug_assert!(intermediate_steps.last().unwrap().start_percentage == 1.);
+
+        intermediate_steps
+    }
+
+    fn update_from_step(
+        &mut self,
+        step: &KeyframesStep,
+        context: &SharedStyleContext,
+        base_style: &Arc<ComputedValues>,
```
Comment (emilio) [inline components/style/animation.rs:227]: I think this can just be `&ComputedValues`, and same in the caller.

**Review by emilio — APPROVED** [review]
Yeah, nice progression :)

File: components/style/animation.rs [hunk 3]
```diff
@@ -553,58 +677,46 @@ impl Animation {
...
         let mut new_style = (**style).clone();
-        let mut update_style_for_longhand = |longhand| {
-            let from = AnimationValue::from_computed_values(longhand, &prev_keyframe_style)?;
-            let to = AnimationValue::from_computed_values(longhand, &next_keyframe_style)?;
+        for (index, from) in prev_keyframe.values.iter().enumerate() {
             PropertyAnimation {
-                from,
-                to,
+                from: from.clone(),
+                to: next_keyframe.values[index].clone(),
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply components/style/animation.rs]: Your suggestion is a lot clearer.

File: components/style/animation.rs (hunk 1 above)
Comment (mrobinson) [reply components/style/animation.rs]: :+1:

File: components/style/animation.rs (hunk 2 above)
Comment (mrobinson) [reply components/style/animation.rs]: Oh, nice.

**Comment by mrobinson** [comment]
@bors-servo r=jdm,emilio

Thank you both for the reviews!

**Pull request merged**